File: src/com/androhi/androiddrawableviewer/form/SettingsDialog.java
package com.androhi.androiddrawableviewer.form;

import com.androhi.androiddrawableviewer.Constants;
import com.androhi.androiddrawableviewer.PluginConfig;
import com.intellij.openapi.fileChooser.FileChooserDescriptor;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.DialogWrapper;
import com.intellij.openapi.ui.TextBrowseFolderListener;
import com.intellij.openapi.ui.TextFieldWithBrowseButton;
import com.intellij.openapi.ui.ValidationInfo;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.openapi.vfs.VirtualFileManager;
import com.intellij.openapi.wm.ToolWindowManager;
import com.intellij.ui.content.Content;
import com.intellij.ui.content.ContentManager;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;

public class SettingsDialog extends DialogWrapper {

    private Project project;
    private JPanel mainPanel;
    private TextFieldWithBrowseButton resDirText;
    private PluginConfig pluginConfig;

    public SettingsDialog(Project project) {
        super(project, true);

        this.project = project;
        setTitle("Settings");
        setResizable(true);

        buildViews();
        init();
    }

    private void buildViews() {
        pluginConfig = PluginConfig.getInstance(project);
        if (project == null || pluginConfig == null) return;

        String savedResDir = pluginConfig.getResDir();
        if (savedResDir == null) {
            savedResDir = project.getBasePath() + Constants.DEFAULT_RESOURCE_PATH;
        }

        resDirText.setText(savedResDir);
        VirtualFile selectDir = VirtualFileManager.getInstance().findFileByUrl(savedResDir);
        FileChooserDescriptor descriptor = new FileChooserDescriptor(false, true, false, false, false, false);
        descriptor.setRoots(selectDir);
        resDirText.addBrowseFolderListener(new TextBrowseFolderListener(descriptor, project));
    }

    @Nullable
    @Override
    protected JComponent createCenterPanel() {
        return mainPanel;
    }

    @Nullable
    @Override
    protected ValidationInfo doValidate() {
        String resDirString = resDirText.getText();
        if (resDirString == null || resDirString.isEmpty()) {
            return new ValidationInfo("Select resource directory.");
        }
        return null;
    }

    @Override
    protected void doOKAction() {
        super.doOKAction();
        String resDirString = resDirText.getText();
        pluginConfig.setResDir(resDirString);
        resetContent(project);
    }

    private void resetContent(Project project) {
        DrawableViewer drawableViewer = new DrawableViewer(project);
        ContentManager contentManager = ToolWindowManager.getInstance(project)
                .getToolWindow(Constants.TOOL_WINDOW_ID).getContentManager();
        Content content = contentManager.getFactory().createContent(drawableViewer, null, false);

        contentManager.removeAllContents(true);
        contentManager.addContent(content);
    }
}

File: src/main/kotlin/com/androhi/androiddrawableviewer/form/SettingsDialog.kt
package com.androhi.androiddrawableviewer.form

import com.androhi.androiddrawableviewer.Constants
import com.androhi.androiddrawableviewer.PluginConfig
import com.intellij.openapi.fileChooser.FileChooserDescriptor
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.DialogWrapper
import com.intellij.openapi.ui.TextBrowseFolderListener
import com.intellij.openapi.ui.TextFieldWithBrowseButton
import com.intellij.openapi.ui.ValidationInfo
import com.intellij.openapi.vfs.VirtualFileManager
import com.intellij.openapi.wm.ToolWindowManager
import javax.swing.JCheckBox
import javax.swing.JComponent
import javax.swing.JPanel

class SettingsDialog(private val project: Project?) : DialogWrapper(project, true) {

    private var mainPanel: JPanel? = null
    private var srcDirText: TextFieldWithBrowseButton? = null
    private var pluginConfig: PluginConfig? = null

    private var checkDrawableMdpi: JCheckBox? = null
    private var checkDrawableHdpi: JCheckBox? = null
    private var checkDrawableXhdpi: JCheckBox? = null
    private var checkDrawableXxhdpi: JCheckBox? = null
    private var checkDrawableXxxhdpi: JCheckBox? = null

    private var checkMipmapMdpi: JCheckBox? = null
    private var checkMipmapHdpi: JCheckBox? = null
    private var checkMipmapXhdpi: JCheckBox? = null
    private var checkMipmapXxhdpi: JCheckBox? = null
    private var checkMipmapXxxhdpi: JCheckBox? = null

    init {
        title = "Settings"
        setResizable(true)

        buildViews()
        init()
    }

    private fun buildViews() {
        if (project == null) return

        pluginConfig = PluginConfig.getInstance(project)

        var savedResDir = pluginConfig?.resDir
        if (savedResDir == null) {
            savedResDir = project.basePath + Constants.DEFAULT_SOURCE_PATH
        }
        srcDirText?.text = savedResDir

        val fileChooserDescriptor = FileChooserDescriptor(false, true, false, false, false, false)
        val selectDir = VirtualFileManager.getInstance().findFileByUrl(savedResDir)
        selectDir?.let {
            fileChooserDescriptor.setRoots(selectDir)
        }
        srcDirText?.addBrowseFolderListener(TextBrowseFolderListener(fileChooserDescriptor, project))
        
        checkDrawableMdpi?.isSelected = pluginConfig?.isDrawableMdpi == true
        checkDrawableHdpi?.isSelected = pluginConfig?.isDrawableHdpi == true
        checkDrawableXhdpi?.isSelected = pluginConfig?.isDrawableXhdpi == true
        checkDrawableXxhdpi?.isSelected = pluginConfig?.isDrawableXxhdpi == true
        checkDrawableXxxhdpi?.isSelected = pluginConfig?.isDrawableXxxhdpi == true

        checkMipmapMdpi?.isSelected = pluginConfig?.isMipmapMdpi == true
        checkMipmapHdpi?.isSelected = pluginConfig?.isMipmapHdpi == true
        checkMipmapXhdpi?.isSelected = pluginConfig?.isMipmapXhdpi == true
        checkMipmapXxhdpi?.isSelected = pluginConfig?.isMipmapXxhdpi == true
        checkMipmapXxxhdpi?.isSelected = pluginConfig?.isMipmapXxxhdpi == true
    }

    override fun createCenterPanel(): JComponent? = mainPanel

    override fun doValidate(): ValidationInfo? {
        val srcDir = srcDirText?.text
        if (srcDir.isNullOrEmpty()) {
            return ValidationInfo("Select source directory.")
        }

        if (hasUncheckedDir()) {
            return ValidationInfo("Check any box.")
        }

        return null
    }

    private fun hasUncheckedDir(): Boolean =
            checkDrawableMdpi?.isSelected != true && checkDrawableHdpi?.isSelected != true &&
                checkDrawableXhdpi?.isSelected != true && checkDrawableXxhdpi?.isSelected != true &&
                checkDrawableXxxhdpi?.isSelected != true

    override fun doOKAction() {
        super.doOKAction()

        pluginConfig?.let {
            it.resDir = null
            it.srcDir = srcDirText?.text

            it.isDrawableMdpi = checkDrawableMdpi?.isSelected == true
            it.isDrawableHdpi = checkDrawableHdpi?.isSelected == true
            it.isDrawableXhdpi = checkDrawableXhdpi?.isSelected == true
            it.isDrawableXxhdpi = checkDrawableXxhdpi?.isSelected == true
            it.isDrawableXxxhdpi = checkDrawableXxxhdpi?.isSelected == true

            it.isMipmapMdpi = checkMipmapMdpi?.isSelected == true
            it.isMipmapHdpi = checkMipmapHdpi?.isSelected == true
            it.isMipmapXhdpi = checkMipmapXhdpi?.isSelected == true
            it.isMipmapXxhdpi = checkMipmapXxhdpi?.isSelected == true
            it.isMipmapXxxhdpi = checkMipmapXxxhdpi?.isSelected == true
        }

        resetContent()
    }

    private fun resetContent() {
        project?.let {
            val drawableViewer = DrawableViewer(it)
            ToolWindowManager.getInstance(it).getToolWindow(Constants.TOOL_WINDOW_ID).contentManager.apply {
                val content = factory.createContent(drawableViewer, null, false)
                removeAllContents(true)
                addContent(content)
            }
        }
    }
}